### DISCUSSION TITLE
Can this be used to dump new G-code?

### DISCUSSION THREAD

**Discussion by prprenti** [Willmac16/OctoPrint-GcodeLeveling#16] [Q&A]
I have a rather odd question/request and hoping maybe you can point me in the right direction.

As I understand, your plugin takes XYZ movements along with a surface mesh and adjusts the Z-heights accordingly.  Is it possible to use your plug to read in gcode, make the adjustments and write out new g-code?

Some background:

I'm a custom cue maker and often work with extremely small/detailed designs that need to have pockets milled out.  My 4th axis is mounted to the bed of my CNC router, however due to the nature taper of the cue, Z0 at X0 is different than Z0 at Xn.  The difference is basically a 0.4 degree line.  Butt end of cue is 1.25" and the joint end is 0.840" over 29".  The equals a 0.4 degree taper.

I want to be able to transform the G-code so my pocket Z-depths are the same.  Essentially what I'm after is a way to "tilt" the entire G-code to a plane parallel to the taper angle of the cue.  I think your software is doing exactly what I'm after, but is doing it on the fly and sending the data directly to the printer.

![image](https://user-images.githubusercontent.com/94191617/177807239-102d422d-e20c-43ca-ade1-3621633fc396.png)

An example is the white rectangle pictured above.  My CAD/CAM software is setup such that the rectangle pocket will be cut into a flat stock at .120" deep.  The problem is the left side of the rectangle Z0 is higher than the right side.  So the left side is .120" deep but the other end may only be 0.65" deep.  If the software was aware of a surface bed mesh of 0.4 degree taper angle, it could offset the Z-heights accordingly and give me a pocket that is consistent depth.

I hope this is making sense, if not I can try and explain in more detail.

Thanks,

Phil

**Comment by Willmac16** [comment]
Yeah this sounds pretty cool and should be doable with the plugin. You will wanna be very careful and incrementally test possibly one line at a time since I wouldn't want the toolhead to crash into the fourth axis.

Config wise: you'll want to have a polynomial/model degree of 1 for the x axis and 0 for the y (i.e. (1, 0)) so you only get a straight taper along the cue.

You could measure as few as two points along taper and enter them into the plugin settings; however, grabbing a few more along the taper couldn't hurt. You'll probably have to do this by carefully lowering the tool and sliding a sheet of paper under to tell when the tool is about to touch, since I don't think any auto probes would work.

Let me know how it goes and I'd be happy to answer any questions that come up.

**Reply by prprenti** [reply]
Is there a way to run this code without OctoPrint?

**Reply by Willmac16** [reply]
Currently there isn't. I'm working full time right now so spinning up a command line interface wouldn't be super quick/easy to fit into my schedule. However, if you're willing to wait maybe a week or two, I should be able to get that done.

**Reply by prprenti** [reply]
Yeah waiting a few weeks is not a problem at all :-)

**Reply by Willmac16** [reply]
Run this to install directly from the GitHub (I don't plan on publishing this to PyPI)
`python3 -m pip install git+https://github.com/Willmac16/OctoPrint-GcodeLeveling.git@standalone `
The standalone can be run from the command line with
`gcode-level`
it has its own help text but here's a rich text guide to using it

* Run `gcode-level -g` once to generate a default config
    - the config file will live in the folder you run this command from, so make sure to run it somewhere where you want the config to live
* edit `leveling-config.yaml` with your favorite text editor (e.g. atom, vscode, or vim)
    - the options are named about the same as they are in the plugin version
    - your config should look something like
         + please make sure the config matches how your work-coords on your machine work
         + since you're only asking for a tilted plane the XY offset doesn't matter a ton, but matching up the z offset is ideal so my code can check for an excessively low position might help catch a dangerous move
    ```
    deg:
    - 1
    - 0
    maxArc: 1.0
    maxLine: 10.0
    points:
    - - 0.0
      - 0.0
      - 0.0
    - - 29.0
      - 0.0
      - -0.41
    zMax: 20.0
    zMin: -0.625
    ```
* run the code on a file `gcode-level path_to_your_toolpath.gcode -c path_to_your_config.yaml
* A new file should be saved with the transformation applied
* CAREFULLY run the transformed gcode one line at a time
    - I can't personally test my program on your machine so its very possible that my code and your flavor of gcode don't play well together
    - I really don't want you to run into a crash so I would suggest running
        + the code inside your CAM software simulation first (there are often manual toolpath features for CAM software)
        + without anything in the machine
        + at low feed rates
        + one line at a time
    - Once you have validated that the code is playing nice, then you could try cutting actual material at speed
+ Let me know how things go, but please be careful. My code only translates a file one line at a time; it doesn't have the proper collision detection abilities of CAM software

**Reply by Willmac16** [reply]
and please don't hesitate to ask any questions that come up

**Reply by Willmac16** [reply]
@prprenti Just wanted to ping you in case you didn't get notified when I posted this a week ago
